openstack/cdn/v1/services/requests.go

```diff
@@ -0,0 +1,319 @@
+
+	return s, nil
+}
+
+// Create accepts a CreateOpts struct and creates a new CDN service using the
+// values provided.
+func Create(c *gophercloud.ServiceClient, opts CreateOptsBuilder) CreateResult {
+	var res CreateResult
+
+	reqBody, err := opts.ToCDNServiceCreateMap()
+	if err != nil {
+		res.Err = err
+		return res
+	}
+
+	// Send request to API
+	resp, err := perigee.Request("POST", createURL(c), perigee.Options{
+		MoreHeaders: c.AuthenticatedHeaders(),
+		ReqBody:     &reqBody,
+		OkCodes:     []int{202},
+	})
+	res.Header = resp.HttpResponse.Header
+	res.Err = err
+	return res
+}
+
+// Get retrieves a specific service based on its URL or its unique ID. For
+// example, both "96737ae3-cfc1-4c72-be88-5d0e7cc9a3f0" and
+// "https://global.cdn.api.rackspacecloud.com/v1.0/services/96737ae3-cfc1-4c72-be88-5d0e7cc9a3f0"
+// are valid options for idOrURL.
+func Get(c *gophercloud.ServiceClient, idOrURL string) GetResult {
+	var url string
+	if strings.Contains(idOrURL, "/") {
+		url = idOrURL
+	} else {
+		url = getURL(c, idOrURL)
+	}
+
+	var res GetResult
+	_, res.Err = perigee.Request("GET", url, perigee.Options{
+		MoreHeaders: c.AuthenticatedHeaders(),
+		Results:     &res.Body,
+		OkCodes:     []int{200},
+	})
+	return res
+}
+
+// UpdateOptsBuilder is the interface options structs have to satisfy in order
+// to be used in the main Update operation in this package. Since many
+// extensions decorate or modify the common logic, it is useful for them to
+// satisfy a basic interface in order for them to be used.
+type UpdateOptsBuilder interface {
+	ToCDNServiceUpdateMap() ([]map[string]interface{}, error)
+}
+
+// Op represents an update operation.
+type Op string
+
+var (
+	// Add is a constant used for performing a "add" operation when updating.
+	Add Op = "add"
+	// Remove is a constant used for performing a "remove" operation when updating.
+	Remove Op = "remove"
+	// Replace is a constant used for performing a "replace" operation when updating.
+	Replace Op = "replace"
+)
+
+// UpdateOpts represents the attributes used when updating an existing CDN service.
+type UpdateOpts []UpdateOpt
+
+// UpdateOpt represents a single update to an existing service. Multiple updates
+// to a service can be submitted at the same time. See UpdateOpts.
+type UpdateOpt struct {
+	// Specifies the update operation to perform.
+	Op Op `json:"op"`
+	// Specifies the JSON Pointer location within the service's JSON representation
+	// of the service parameter being added, replaced or removed.
+	Path string `json:"path"`
+	// Specifies the actual value to be added or replaced. It is not required for
+	// the remove operation.
+	Value map[string]interface{} `json:"value,omitempty"`
+}
+
+// ToCDNServiceUpdateMap casts an UpdateOpts struct to a map.
+func (opts UpdateOpts) ToCDNServiceUpdateMap() ([]map[string]interface{}, error) {
+	s := make([]map[string]interface{}, len(opts))
+
+	for i, opt := range opts {
+		if opt.Op == "" {
+			return nil, no("Op")
```

Comment:
You should also validate that `opt.Op` is one of `Add`, `Remove` or `Replace` here, because technically a client can cast any string to an `Op`.

The *other* option is to change `type Op string` to `type Op struct { name string }`, which prevents casting, but would require slightly more code to pull the `name` field out here - your choice.